### PR TITLE
[Filters] Refactor to use new FiltersBar component

### PR DESCRIPTION
**Pull request by mrcthms**
### WHY are these changes introduced?

This PR takes another approach to the applied filters and pinned filters logic. In [this PR](https://github.com/Shopify/polaris/pull/10566), we tried to solve for an issue on a specific index table, the Orders index, which loads the filter data asynchronously. However, I now feel this approach is wrong and we should be solving for this problem in another way.

Due to how the Filters component was built before, all the logic regarding state management of pinned and applied filters was generated when the Filters component mounted, yet if there were no filters array passed to the `Filters` component, it would not render. This caused issues around keeping applied filters and pinned filters in sync with each other.

The component would mount with an empty `filters` and `appliedFilters` array, and then re-render once the data was available and passed down. We should not be invoking any of the internal state logic around pinned/applied filters until we have a list of filters to parse. 

This PR breaks out the Filters component, and puts all the stateful logic inside a new FiltersBar child component of the Filters, and this only gets rendered when `filters.length > 0`. This allows us to keep the initial pinned/applied filter state in sync and prevents and issues we were seeing around applied filters jumping around the list.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

Spin URL: https://admin.web.filters-fix.marc-thomas.eu.spin.dev/store/shop1/orders?inContextTimeframe=none&delivery_method=shipping

- Confirm that the Delivery method filter pill appears when the data has loaded
- Apply more filters, and then reload the page
- Confirm that those filters appear when the data has loaded
- Confirm that filters stay in their existing position, until deleted.

Spin URL: https://admin.web.filters-fix.marc-thomas.eu.spin.dev/store/shop1/products?selectedView=all

- Apply a vendor filter to the index table
- Confirm that the Vendor filter pill remains in the same place when a value is selected

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Review by ouellettejordan — APPROVED** [review]
Tested spin link and looks good to me 👍

File: polaris-react/src/utilities/index-filters/IndexFiltersManager.tsx [hunk 1]
```diff
@@ -10,6 +10,7 @@ export interface IndexFiltersManagerProps {
 
 type Context = NonNullable<ContextType<typeof IndexFiltersModeContext>>;
 
+/** @deprecated No longer needed, will be removing in v12 */
```
Comment (sam-b-rose) [inline polaris-react/src/utilities/index-filters/IndexFiltersManager.tsx]: Should this deprecation be merged into `main` instead of `12.x.x`?

**Review by sam-b-rose — APPROVED** [review]
Changes look good! Should we move the deprecation notice to a separate PR targeting the `main` branch, then make the v12 updates targeting `12.x.x`?

**Pull request merged**